sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1430,7 +1430,23 @@ class Dataset[T] private[sql](
    */
   @scala.annotation.varargs
   def select(cols: Column*): DataFrame = withPlan {
-    Project(cols.map(_.named), logicalPlan)
+    val untypedCols = cols.map {
+      case typedCol: TypedColumn[_, _] =>
+        val isSimpleEncoder = typedCol.encoder.namedExpressions.head match {
+          case Alias(_: BoundReference, _) if !typedCol.encoder.isSerializedAsStruct => true
+          case _ => false
+        }
+        if (isSimpleEncoder) {
+          // This typed column produces simple type output that can be fit into untyped `DataFrame`.
+          typedCol.withInputType(exprEnc, logicalPlan.output)
```

Comment:
Previously we didn't call `withInputType` for `count`, right?

Reply:
`count` has no `TypedAggregateExpression`. `withInputType` only works on `TypedAggregateExpression`.

Reply:
I mean, `df.select(count("*"))` works without calling `withInputType`, right?

Reply:
Yes, for `TypedColumn` that doesn't contain `TypedAggregateExpression`, `withInputType` is no-op, so you don't need to call `withInputType` for `df.select(count("*"))`.

Reply:
So here we are supporting more cases than before?

Reply:
Oh, I get your point now. Yea, we should not allow more cases than before.